### ISSUE TITLE
[BUG] Premature __deinit__ UAF: raw pointer escaping an owned local reads freed memory (1.0.0 stable)

### ISSUE THREAD

**Issue by mvillmow** [modular/mojo#6963]
### Summary

On Mojo 1.0.0 stable, a struct that owns heap storage and exposes a method returning a raw
`Pointer` into that storage will have its `__deinit__` hoisted to right after the **last
syntactic use of the struct** in the caller. If the caller holds the returned pointer in a
local and keeps using it afterwards (the standard "capture `data_ptr()` then loop" pattern),
the storage is freed mid-use → use-after-free. The same code shape passes on 1.0.0b2.

This is the same class as #6959 (closed COMPLETED), #6707 (closed NOT_PLANNED) and #6939
(closed DUPLICATE), but with a much more common trigger shape: any struct with
refcounted-or-not heap storage and a `data_ptr`-style accessor. Our codebase has ~220 call
sites in this shape (tensor kernels holding `_data` in locals) and a confirmed live case.

### Minimal reproducer (stdlib only)

```mojo
"""Self-contained: run with `mojo run repro.mojo` on 1.0.0 stable."""
from std.memory import Pointer
from std.memory.alloc import unsafe_alloc


struct MiniTensor:
    """Minimal refcounted tensor with an escaping raw data_ptr()."""

    var _data: Pointer[Float32, MutUntrackedOrigin]
    var _refcount: Pointer[Int, MutUntrackedOrigin]
    var _n: Int

    def __init__(out self, n: Int):
        self._n = n
        self._data = unsafe_alloc[Float32](n)
        self._refcount = unsafe_alloc[Int](1)
        self._refcount[] = 1

    def __init__(out self, *, copy: Self):
        self._data = copy._data
        self._refcount = copy._refcount
        self._refcount[] += 1
        self._n = copy._n

    def __init__(out self, *, deinit move: Self):
        self._data = move._data
        self._refcount = move._refcount
        self._n = move._n

    def data_ptr(self) -> Pointer[Float32, MutUntrackedOrigin]:
        """Escape: returns an untracked raw pointer into our storage."""
        return self._data

    def set(self, i: Int, v: Float32):
        self._data[unsafe_offset=i] = v

    def load(self, i: Int) -> Float32:
        return self._data[unsafe_offset=i]

    def __deinit__(deinit self):
        self._refcount[] -= 1
        if self._refcount[] == 0:
            self._data.unsafe_free()
            self._refcount.unsafe_free()


@always_inline
def _opaque(x: Float32) -> Float32:
    """Opaque op the compiler cannot fold away."""
    return x * Float32(3.0) + Float32(1.0)


def main() raises:
    var n = 1024
    var t = MiniTensor(n)
    for i in range(n):
        t.set(i, Float32(i % 7))

    # The escape: last syntactic use of `t` is this call.
    var ptr = t.data_ptr()

    # Loop reads exclusively through the escaped pointer.
    var acc = Float32(0.0)
    for i in range(n):
        acc += _opaque(ptr[unsafe_offset=i])

    var expected = Float32(0.0)
    for i in range(n):
        var v = Float32(i % 7)
        expected += v * Float32(3.0) + Float32(1.0)

    print("acc     =", acc)
    print("expected=", expected)
    if acc != expected:
        raise Error("MISMATCH: escaped data_ptr read freed memory")
    print("OK")
```

### Cross-version evidence (identical logic, 3 runs each)

| Compiler | Result |
| --- | --- |
| **1.0.0 stable** (`ed45d567`) | ❌ **FAIL 3/3** — `acc = 10222.0`, `expected = 10225.0` (3 elements clobbered) |
| **1.0.0b2** (`2cf4d08a`, b2-flavored mirror with `UnsafePointer`/`alloc`/`free`) | ✅ PASS 3/3 |
| **1.0.0 stable control** (same struct, loop reads via `load()` — no escaped pointer) | ✅ PASS 3/3 |

The control passing proves the write path and math are correct; the 3 clobbered elements are
consistent with the freed 4 KiB block being reused (tcmalloc writes its freelist pointer into
the block header). `mojo build --emit llvm` confirms the refcount drop + `AlignedFree` land
before the read loop in the failing case.

### Production impact

This is the `AnyTensor.data_ptr()` / direct `_data` pattern used throughout our ML framework —
most visibly in evaluation/accuracy loops:

```mojo
var logits = model.forward(batch_images)
var logits_data = logits.data_ptr[DType.float32]()   # last use of logits
for i in range(current_batch_size):
    var max_logit = logits_data[unsafe_offset=i * num_classes]   # reads freed memory
```

Kernel functions taking tensors as **parameters** are unaffected (the owner lives in the
caller frame, so the callee cannot have the deinit hoisted) — the trigger requires an
**owned local** whose accessor call is its last syntactic use. It currently passes in our
suite only by heap/codegen luck; a working probe (`repro/repro_tensor_inline.mojo` in
HomericIntelligence/Odyssey, branch `5800-mojo-1.0.0`) fails deterministically.

### Known workaround (from #6707)

Tying the returned pointer's origin to the owner fixes it (validated on the atomic shape in
#6959). The returned pointer must be `origin_of(self)`-tracked via `unsafe_origin_cast`:

```mojo
def data_ptr(mut self) -> Pointer[Float32, origin_of(self)]:
    return self._data.unsafe_origin_cast[origin_of(self)]()
```

This requires the accessor to be `mut self` and cascades `mut` through callers — which is why
we're reporting rather than silently changing a public API.

### Related

- #6959 (same class, closed COMPLETED — atomic accessor shape; this issue is the tensor
  `data_ptr` shape, still failing on stable)
- #6707 (origin-tie workaround)
- #6939 (return-move deinit UAF, closed DUPLICATE)

**Comment by mvillmow** [comment]
Hi @dgurchenkov — confirming this is a **regression from `1.0.0b2`**: the self-contained repro **passes 3/3 on `mojo==1.0.0b2` (2cf4d08a) and fails 3/3 on `1.0.0` stable (ed45d567)**.

Cross-version evidence (HomericIntelligence/Odyssey, branch `5800-mojo-1.0.0`):
- `repro/repro_tensor_inline.mojo` — stable, FAIL 3/3 (`acc = 10222.0`, expected `10225.0`)
- `repro/repro_tensor_inline_b2.mojo` — b2 mirror, PASS 3/3
- `repro/repro_tensor_inline_control.mojo` — stable control (no escape), PASS 3/3

We've applied the #6707 origin-tie workaround (`AnyTensor.data_ptr` now returns `Pointer[..., origin_of(self)]` via `unsafe_origin_cast`), which eliminates the failure, but the underlying compiler behavior (hoisting `__deinit__` when a raw pointer escapes an owned local) still needs an upstream fix — this shape is pervasive (~220 `_data` escape sites in our tensor kernels), so a real fix matters for the ecosystem.

**Comment by NathanSWard** [comment]
@mvillmow - this is intended behavior and not a bug.
This is a duplicate of: https://github.com/modular/modular/issues/6707#issuecomment-4858172977

We have a collection of good learning materials for the Mojo origin system, lifetimes, and pointer interaction here: https://mojolang.org/docs/manual/values/lifetimes/

I would recommend going through these to understand _why_ the issue you encountered happened, and how the correct fix is to use proper origin-tracking.

If you need help brainstorming ideas on how to better express this without running into memory issues you're always welcome to post on our forum and discord to get insight on how to better restructure your code 😄. That being said, we'd also prefer to have this discussion with a person and not asynchronously over GitHub issue via an LLM.

**Comment by mvillmow** [comment]
@NathanSWard I've read the duplicate bug and the lifetime. This looks like a broken compiler optimization. If the compiler cannot determine the final use, the final use is at end of the scope. Inserting a destruction of the object at any point before the end of the scope is an optimization, so unless it can be proven that it is safe, the compiler needs to be conservative and place it at the end of the scope. This code is really no different than using std::vector.data() in C++ and then processing the raw data directly instead of using the vector interfaces. Are you telling me that using std::vector.data() style coding is invalid in Mojo? I get its unsafe, hence the tagging it as unsafe and untracked, so trust that I'm doing things correctly, but invalid seems to be overkill.
